x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/util/BatchedDataIterator.java

```diff
@@ -0,0 +1,189 @@
+        this.count = 0;
+        this.isScrollInitialised = false;
+    }
+
+    /**
+     * Returns {@code true} if the iteration has more elements.
+     * (In other words, returns {@code true} if {@link #next} would
+     * return an element rather than throwing an exception.)
+     *
+     * @return {@code true} if the iteration has more elements
+     */
+    public boolean hasNext() {
+        return !isScrollInitialised || count != totalHits;
+    }
+
+    /**
+     * The first time next() is called, the search will be performed and the first
+     * batch will be given to the listener. Any subsequent call will return the following batches.
+     * <p>
+     * Note that in some implementations it is possible that when there are no
+     * results at all, the first time this method is called an empty Collection is given to the listener.
+     */
+    public void next(ActionListener<E> listener) {
+        if (!hasNext()) {
+            listener.onFailure(new NoSuchElementException());
+        }
+        ActionListener<SearchResponse> wrappedListener = ActionListener.wrap(
+            searchResponse -> {
+                scrollId = searchResponse.getScrollId();
+                mapHits(searchResponse, listener);
+            },
+            listener::onFailure
+        );
+        if (scrollId == null) {
+            initScroll(wrappedListener);
+        } else {
+            SearchScrollRequest searchScrollRequest = new SearchScrollRequest(scrollId).scroll(CONTEXT_ALIVE_DURATION);
+            ClientHelper.executeAsyncWithOrigin(client.threadPool().getThreadContext(),
+                ClientHelper.DATA_FRAME_ORIGIN,
+                searchScrollRequest,
+                wrappedListener,
+                client::searchScroll);
+        }
+    }
+
+    private void initScroll(ActionListener<SearchResponse> listener) {
+        LOGGER.trace("ES API CALL: search index {}", index);
+
+        isScrollInitialised = true;
+
+        IndicesOptions indicesOptions = SearchRequest.DEFAULT_INDICES_OPTIONS;
+        indicesOptions = IndicesOptions.fromOptions(true, indicesOptions.allowNoIndices(), indicesOptions.expandWildcardsOpen(),
+            indicesOptions.expandWildcardsClosed(), indicesOptions);
+
+        SearchRequest searchRequest = new SearchRequest(index);
+        searchRequest.indicesOptions(indicesOptions);
+        searchRequest.scroll(CONTEXT_ALIVE_DURATION);
+        searchRequest.source(new SearchSourceBuilder()
+            .fetchSource(getFetchSourceContext())
+            .size(BATCH_SIZE)
+            .query(getQuery())
+            .trackTotalHits(true)
+            .sort(sortField(), sortOrder()));
+
+
+        ActionListener<SearchResponse> wrappedListener = ActionListener.wrap(
+            searchResponse -> {
+                totalHits = searchResponse.getHits().getTotalHits().value;
+                scrollId = searchResponse.getScrollId();
+                listener.onResponse(searchResponse);
```

Comment:
Set `isScrollInitialised = true;` here or make it a function of `scrollId`

In fact I think it would be more consistent to remove it in preference to  `if (scrollId == null) ` as is used elsewhere such as `next()` L86

Reply:
nice catch!

Reply:
Though its a good thought, there are apparently situations where the total hits returned will be `0` and the `scrollId` returned will be `null` in those situations, this will never exit :(. So, I opted to go back to `isScrollInitialised`. 

If the `scrollId` is actually `null` but not all hits have been gathered, the scroll request will throw an error and exit, which is acceptable, I think.